File: app/src/main/java/com/jbrunton/mymovies/discover/DiscoverFragment.java
package com.jbrunton.mymovies.discover;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.jbrunton.mymovies.R;
import com.jbrunton.mymovies.search.SearchResultsAdapter;
import com.jbrunton.mymovies.search.SearchViewState;
import com.jbrunton.mymovies.shared.BaseFragment;
import com.jbrunton.mymovies.shared.LoadingStateContext;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class DiscoverFragment extends BaseFragment<DiscoverViewModel> {
    private SearchResultsAdapter nowPlayingAdapter;
    @BindView(R.id.discover_content) View discoverContent;
    @BindView(R.id.now_playing) RecyclerView nowPlayingList;
    private LoadingStateContext loadingStateContext = new LoadingStateContext();

    @Nullable @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_discover, container, false);
        ButterKnife.bind(this, view);
        ButterKnife.bind(loadingStateContext, view);

        nowPlayingList.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.HORIZONTAL, false));
        nowPlayingAdapter = new SearchResultsAdapter(getActivity(), R.layout.item_movie_card_grid);
        nowPlayingList.setAdapter(nowPlayingAdapter);

        return view;
    }

    @Override public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getViewModel().viewState().observe(this, this::updateView);
    }

    @Override protected DiscoverViewModel provideViewModel() {
        DiscoverViewModel.Factory factory = new DiscoverViewModel.Factory(dependencies().getMoviesRepository());
        return getViewModel(DiscoverViewModel.class, factory);
    }

    @OnClick(R.id.genres_link)
    public void onGenresClick() {
        Intent intent = new Intent(getActivity(), GenresActivity.class);
        startActivity(intent);
    }

    private void updateView(SearchViewState viewState) {
        discoverContent.setVisibility(toVisibility(viewState.loadingViewState().showContent()));
        nowPlayingAdapter.setDataSource(viewState.movies());
        loadingStateContext.updateView(viewState.loadingViewState());
    }

    protected int toVisibility(boolean show) {
        return show ? View.VISIBLE : View.GONE;
    }
}

File: app/src/main/java/com/jbrunton/mymovies/discover/DiscoverFragment.kt
package com.jbrunton.mymovies.discover

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import com.jakewharton.rxbinding2.view.clicks
import com.jbrunton.mymovies.R
import com.jbrunton.mymovies.helpers.observe
import com.jbrunton.mymovies.helpers.toVisibility
import com.jbrunton.mymovies.search.SearchResultsAdapter
import com.jbrunton.mymovies.search.SearchViewState
import com.jbrunton.mymovies.shared.BaseFragment
import com.jbrunton.mymovies.shared.LoadingViewState
import com.jbrunton.mymovies.shared.Success
import com.trello.rxlifecycle2.android.lifecycle.kotlin.bindToLifecycle
import kotlinx.android.synthetic.main.fragment_discover.*
import org.koin.androidx.viewmodel.ext.android.viewModel

class DiscoverFragment : BaseFragment<DiscoverViewModel>() {
    private lateinit var nowPlayingAdapter: SearchResultsAdapter

    private val viewModel: DiscoverViewModel by viewModel()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_discover, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        nowPlayingAdapter = SearchResultsAdapter(activity!!, R.layout.item_movie_card_grid)
        now_playing.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(activity, androidx.recyclerview.widget.LinearLayoutManager.HORIZONTAL, false)
        now_playing.adapter = nowPlayingAdapter

        genres_link.clicks()
                .bindToLifecycle(this)
                .subscribe {
                    val intent = Intent(activity, GenresActivity::class.java)
                    startActivity(intent)
                }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel.viewState.observe(this, this::updateView)
        viewModel.start()
    }

    private fun updateView(viewState: LoadingViewState<SearchViewState>) {
        discover_content.visibility = toVisibility(viewState is Success)
        if (viewState is Success) {
            nowPlayingAdapter.setDataSource(viewState.value)
        }
        updateLoadingView(viewState)
    }
}
